src/libs/OptimisticReportNames.ts

```diff
@@ -0,0 +1,332 @@
+    const formula = titleField.defaultValue;
+    const formulaParts = parse(formula);
+
+    // Check if any formula part might be affected by this update
+    const isAffected = formulaParts.some((part) => {
+        if (part.type === FORMULA_PART_TYPES.REPORT) {
+            // Checking if the formula part is affected in this manner works, but it could certainly be more precise.
+            // For example, a policy update only affects the part if the formula in the policy changed, or if the report part references a field on the policy.
+            // However, if we run into performance problems, this would be a good place to optimize.
+            return updateType === 'report' || updateType === 'policy';
+        }
+        if (part.type === FORMULA_PART_TYPES.FIELD) {
+            return updateType === 'report';
+        }
+        return false;
+    });
+
+    if (!isAffected) {
+        Performance.markEnd(CONST.TIMING.COMPUTE_REPORT_NAME);
+        Timing.end(CONST.TIMING.COMPUTE_REPORT_NAME);
+        return null;
+    }
+
+    // Build context with the updated data
+    const updatedReport =
+        updateType === 'report' && targetReport.reportID === getReportIDFromKey(incomingUpdate.key) ? {...targetReport, ...(incomingUpdate.value as Partial<Report>)} : targetReport;
+
+    const updatedPolicy = updateType === 'policy' && targetReport.policyID === getPolicyIDFromKey(incomingUpdate.key) ? {...(policy ?? {}), ...(incomingUpdate.value as Policy)} : policy;
+
+    // Compute the new name
+    const formulaContext: FormulaContext = {
+        report: updatedReport,
+        policy: updatedPolicy,
+    };
+
+    const newName = compute(formula, formulaContext);
+
+    // Only return an update if the name actually changed
+    if (newName && newName !== targetReport.reportName) {
+        Log.info('[OptimisticReportNames] Report name computed', false, {
+            reportID: targetReport.reportID,
+            oldName: targetReport.reportName,
+            newName,
+            formula,
+            updateType,
+            isNewReport: !report,
+        });
+
+        Performance.markEnd(CONST.TIMING.COMPUTE_REPORT_NAME);
+        Timing.end(CONST.TIMING.COMPUTE_REPORT_NAME);
+        return newName;
+    }
+
+    Performance.markEnd(CONST.TIMING.COMPUTE_REPORT_NAME);
+    Timing.end(CONST.TIMING.COMPUTE_REPORT_NAME);
+    return null;
+}
+
+/**
+ * Update optimistic report names based on incoming updates
+ * This is the main middleware function that processes optimistic data
+ */
+function updateOptimisticReportNamesFromUpdates(updates: OnyxUpdate[], context: UpdateContext): OnyxUpdate[] {
+    Performance.markStart(CONST.TIMING.UPDATE_OPTIMISTIC_REPORT_NAMES);
+    Timing.start(CONST.TIMING.UPDATE_OPTIMISTIC_REPORT_NAMES);
+
+    Log.info('[OptimisticReportNames] Processing optimistic updates for report names', false, {
+        updatesCount: updates.length,
+        hasReports: Object.keys(context.allReports).length > 0,
```

Comment:
NAB: I don't think `hasReports` is that useful, because the context will be initialized before getting to this. Also, maybe we can log after checking for the beta. Otherwise the log is kind of meaningless.